xmtp/src/cryptography.rs

```diff
@@ -0,0 +1,80 @@
+use aes::cipher::{KeyIvInit, StreamCipher};
+use aes::Aes256;
+use ethers_core::k256::elliptic_curve::subtle::ConstantTimeEq;
+use hkdf::hmac::{Hmac, Mac};
+use sha2::Sha256;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum EncryptionError {
+    #[error("bad data")]
+    BadKeyOrIv,
+    #[error("unknown error")]
+    Unknown,
+}
+
+#[derive(Debug, Error)]
+pub enum DecryptionError {
+    #[error("bad ciphertext {0}")]
+    BadCiphertext(String),
+    #[error("bad data")]
+    BadKeyOrIv,
+    #[error("unknown error")]
+    Unknown,
+}
+
+pub(crate) fn aes_256_ctr_encrypt(ptext: &[u8], key: &[u8]) -> Result<Vec<u8>, EncryptionError> {
+    let key: [u8; 32] = key.try_into().map_err(|_| EncryptionError::BadKeyOrIv)?;
+
+    let zero_nonce = [0u8; 16];
+    let mut cipher = ctr::Ctr32BE::<Aes256>::new(key[..].into(), zero_nonce[..].into());
+
+    let mut ctext = ptext.to_vec();
+    cipher.apply_keystream(&mut ctext);
+    Ok(ctext)
+}
+
+fn aes_256_ctr_decrypt(ctext: &[u8], key: &[u8]) -> Result<Vec<u8>, DecryptionError> {
+    aes_256_ctr_encrypt(ctext, key).map_err(|e| match e {
```

Comment:
This is what `libsignal` does. I can't quite get my head around why this works...but it does.